### PR TITLE
Support pointers in compiler

### LINKED ISSUE #1032: Support pointers in the compiler
Not a priority for now, but I think we can add support for pointers via array wrappers. We have #1023 and #1025 adding syscalls that return an array of things and these things have different types:
```c#
        public StackItem ToStackItem(ReferenceCounter referenceCounter)
        {
            return new Array(referenceCounter, new StackItem[]
            {
                // Computed properties                                                                                                                                                                                                                                                                                                                                                       
                Hash.ToArray(),

                // Transaction properties                                                                                                                                                                                                                                                                                                                                                    
                (int)Version,
                Nonce,
                Sender.ToArray(),
                SystemFee,
                NetworkFee,
                ValidUntilBlock,
                Script,
            });
        }
```
We absolutely have to match this API at the syscall level, so we will return an array too. But then the question is how to represent that array in Go interop wrappers. In C# it looks like this:
```c#
        [Syscall("System.Blockchain.GetTransaction")]
        public static extern Transaction GetTransaction(byte[] hash);
...
    public class Transaction
    {
        public readonly byte[] Hash;
        public readonly byte Version;
        public readonly uint Nonce;
        public readonly byte[] Sender;
        public readonly long SystemFee;
        public readonly long NetworkFee;
        public readonly uint ValidUntilBlock;
        public readonly byte[] Script;
    }
```
As object is a reference type there it works fine. However, in Go structs are value types and we have to `CONVERT` this array into struct to support an API like this:
```go
func GetTransaction(hash []byte) Transaction
```
where `Transaction` is a struct. We could avoid this conversion if we're to make it look like this:
```go
func GetTransaction(hash []byte) *Transaction
```
returning a pointer to structure that can be represented as an array.

And this approach can be extended easily, any kind of pointer can be represented via an array wrapper.

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1032 .

Things to discuss:
1. Do we need to support taking pointers from the variable? This can be done by using `Array(Struct)` wrapper instead of `Struct -> Array`, but doesn't worth the effort IMO.
2. Do we need to copy struct when passing it as a parameter? This way we become more compatible with the Go spec, but again does it worth the effort? While simple struct copy can be done with `Struct -> Array -> Struct` conversion, this won't copy non-pointer struct fields, so we need to do something different.

### CODE REVIEW

File: pkg/compiler/codegen.go [hunk 1]
```diff
@@ -1325,7 +1338,19 @@ func (c *codegen) convertMap(lit *ast.CompositeLit) {
...
 }
 
-func (c *codegen) convertStruct(lit *ast.CompositeLit) {
+func (c *codegen) getStruct(typ types.Type) (*types.Struct, bool) {
+	switch t := typ.Underlying().(type) {
+	case *types.Struct:
+		return t, true
+	case *types.Pointer:
+		strct, ok := t.Elem().Underlying().(*types.Struct)
+		return strct, ok
```
Comment (AnnaShaleva) [inline pkg/compiler/codegen.go:1357]: Just
```
return t.Elem().Underlying().(*types.Struct)
```
?

File: pkg/compiler/codegen.go [hunk 2]
```diff
@@ -610,6 +610,16 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
...
 		return nil
 
+	case *ast.StarExpr:
+		_, ok := c.getStruct(c.typeOf(n.X))
+		if !ok {
+			c.prog.Err = errors.New("dereferencing is only supported on structs")
+			return nil
+		}
+		ast.Walk(c, n.X)
+		emit.Instruction(c.prog.BinWriter, opcode.CONVERT, []byte{byte(stackitem.StructT)})
```
Comment (AnnaShaleva) [inline pkg/compiler/codegen.go:620]: We have `(c *codegen) emitConvert` for this.

File: pkg/compiler/codegen.go (hunk 1 above)
Comment (fyrchik) [reply pkg/compiler/codegen.go:1357]: No, this way it will assume single value return and it will not compile.

**Comment by roman-khimov** [comment]
It's very limited, but still it's useful. I think we need to update the documentation to set appropriate expectations on the smart contract developer side.

> Do we need to support taking pointers from the variable?

It'd be nice, but it's not worth changing struct representation, at least for now.

> Do we need to copy struct when passing it as a parameter?

This one looks more important to me as at the moment this test fails:
```
func TestAddressOfLiteral(t *testing.T) {
        src := `package foo                                                                                                                                                                                                                                                                                                                                                                  
        type Foo struct { A int }                                                                                                                                                                                                                                                                                                                                                            
        func Main() int {                                                                                                                                                                                                                                                                                                                                                                    
                f := Foo{A: 4}                                                                                                                                                                                                                                                                                                                                                               
                setA(f, 3)                                                                                                                                                                                                                                                                                                                                                                   
                return f.A                                                                                                                                                                                                                                                                                                                                                                   
        }                                                                                                                                                                                                                                                                                                                                                                                    
        func setA(s Foo, a int) { s.A = a }`
        eval(t, src, big.NewInt(4))
}
```
It's a separate bug, though.

**Comment by fyrchik** [comment]
@roman-khimov I first implemented support for a struct copy on parameter passing, but removed because it wasn't full. Restored, but general case still needs to be addressed.

**Comment by fyrchik** [comment]
Actually no, I have found how to do this in general case. Fixed.

**Comment by roman-khimov** [comment]
Nice, but documentation is still not updated.

**Pull request merged**